dtreeviz/models/shadow_decision_tree.py

```diff
@@ -284,7 +292,7 @@ def get_split_node_heights(self, X_train, y_train, nbins) -> Mapping[int, int]:
             # print(f"\tmax={np.max(height_of_bins):2.0f}, heights={list(height_of_bins)}, {len(height_of_bins)} bins")
         return node_heights
 
-    def predict(self, x: np.ndarray) -> Tuple[Number, List]:
+    def predict(self, x: np.ndarray, path_only: bool = False) -> Tuple[Number, List]:
```

Comment:
why add new arg? For use, `path = self.predict(self.x_data[i], path_only=True)` just do `_, path = self.predict(self.x_data[i])`

Reply:
if I don't send path_only arg, we will receive "RecursionError: maximum recursion depth exceeded while decoding a JSON array from a unicode string" for LightGBM.
This is because get_node_samples() call the predict() method and leaf.prediction() (from inside predict()) will also call for get_node_samples to calculate the leaf prediction.

Reply:
Hmm...ok, let me think about this. the arg seems like a kludge/hack.

Reply:
maybe having different methods for prediction and path ?

Reply:
Why does get node samples call predict?

Reply:
Is there no way to find samples w/o walking the tree very time? i see you caching them but maybe we compute this once up front for any tree.

Reply:
predict() by default returned both prediction and prediction path nodes. To calculate the node samples, for lightgbm, we need to call predict() to get the nodes for each sample prediction. In this way we can calculate the samples for each node for lightgbm.

the challange is that the other libraries have different implementation to get the node samples:
- for sklearn, we can get the path directly for its API, dec_paths = self.tree_model.decision_path(self.x_data)  
- for spark, the leaf predictions is stored in tree metadata, so we dont need to re-calculate the leaf prediction from its node samples (like for lightgbm)

An alternative would be to have two separate functions, one for prediction and one for prediction path, but we will double the computational cost, which can be significant when we have a large number of samples (training)

I think the confusion is also caused by the fact the name of the method is predict() and in lightgbm case it will return only the prediction path.

Reply:
@parrt would be a solution to handle this design issue in another PR ? (because it is related not only to lightgbm).
 
in the end, having two separate functions for prediction and paths, it will not increase the computational cost significantly, because we don't need both of values in costly functions. 
For example, the get_node_samples() is very costly because it handles all the training set examples, but it needs only the paths. The other functions which need both values are called for just one training example.

Reply:
@parrt any thoughts on this ? :)

Reply:
I agree let's merge and then please do create another PR and let's separate the function into two.